File: 2021/day_11/solution.rb
def neighbours(x, y, width, height)
  nb = []
  nb << [x-1, y] if x > 0
  nb << [x, y-1] if y > 0
  nb << [x+1, y] if x < (width - 1)
  nb << [x, y+1] if y < (height - 1)
  nb << [x+1, y+1] if y < (height - 1) && x < (width - 1)
  nb << [x-1, y+1] if y < (height - 1) && x > 0
  nb << [x-1, y-1] if y > 0 && x > 0
  nb << [x+1, y-1] if y > 0 && x < (width - 1)
  nb
end

def flash(x, y, map, flashed)
  height = map.size
  width = map[0].size
  return unless flashed[[x, y]].nil?

  val = map[y][x]
  if val < 9
    map[y][x] = val + 1
  else 
    map[y][x] = 0
    flashed[[x,y]] = true
    neighbours(x, y, width, height).each do |nb|
      flash(nb[0], nb[1], map, flashed)
    end
  end
end

def simulate(map)
  height = map.size
  width = map[0].size
  flashed = {}

  height.times do |y|
    width.times do |x|
      flash(x, y, map, flashed)
    end
  end

  flashed
end

def solve_p1
  map = File.read(ARGV[0]).split("\n").map{|l| l.chars.map(&:to_i) }
  flashes = 0
  100.times do |i|
    flashes += simulate(map).size
  end
  
  puts "Part 1: #{flashes}"
end

def solve_p2
  map = File.read(ARGV[0]).split("\n").map{|l| l.chars.map(&:to_i) }
  steps = 1
  octo_count = map[0].size * map.size
  while simulate(map).size != octo_count do
    steps += 1
  end
  
  puts "Part 2: #{steps}"
end

solve_p1
solve_p2

File: 2021/day_11/solution.rs
use std::collections::HashSet;
use std::env;
use std::fs;

fn neighbours(x: usize, y: usize, width: usize, height: usize) -> Vec<(usize, usize)> {
    let mut nb = Vec::new();

    if x > 0 {
        nb.push((x - 1, y));
    }
    if y > 0 {
        nb.push((x, y - 1));
    }
    if x < (width - 1) {
        nb.push((x + 1, y));
    }
    if y < (height - 1) {
        nb.push((x, y + 1));
    }
    if y < (height - 1) && x < (width - 1) {
        nb.push((x + 1, y + 1));
    }
    if y < (height - 1) && x > 0 {
        nb.push((x - 1, y + 1));
    }
    if y > 0 && x > 0 {
        nb.push((x - 1, y - 1));
    }
    if y > 0 && x < (width - 1) {
        nb.push((x + 1, y - 1));
    }

    nb
}

fn flash(x: usize, y: usize, mut map: &mut Vec<Vec<u32>>, mut flashed: &mut HashSet<(usize, usize)>) {
    let height = map.len();
    let width = map[0].len();
    if flashed.contains(&(x, y)) {
        return;
    }

    let val = map[y][x];
    if val < 9 {
        map[y][x] = val + 1
    } else {
        map[y][x] = 0;
        flashed.insert((x, y));
        for (nbx, nby) in neighbours(x, y, width, height).iter() {
            flash(*nbx, *nby, &mut map, &mut flashed);
        }
    }
}

fn simulate(mut map: &mut Vec<Vec<u32>>) -> usize {
    let height = map.len();
    let width = map[0].len();
    let mut flashed: HashSet<(usize, usize)> = HashSet::new();

    for y in 0..height {
        for x in 0..width {
            flash(x, y, &mut map, &mut flashed);
        }
    }

    flashed.len()
}

fn part_one(mut map: Vec<Vec<u32>>) {
    let mut flashed = 0;
    for _ in 0..100 {
        flashed += simulate(&mut map);
    }

    println!("Part 1: {}", flashed);
}

fn part_two(mut map: Vec<Vec<u32>>) {
  let mut steps = 1;
  let octo_count = map[0].len() * map.len();
  while simulate(&mut map) != octo_count {
    steps += 1
  }
  
  println!("Part 2: {}", steps);

}

fn main() {
    let filename = env::args().nth(1).expect("Please provide input file");
    let map: Vec<Vec<u32>> = fs::read_to_string(filename)
        .unwrap()
        .split("\n")
        .collect::<Vec<_>>()
        .iter()
        .filter(|line| line.len() > 0)
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).unwrap())
                .collect::<Vec<u32>>()
        })
        .collect();

    part_one(map.clone());
    part_two(map.clone());
}
